sites/all/modules/ding_unilogin/ding_unilogin.module

```diff
@@ -0,0 +1,342 @@
+                 $unilogin_id,
+                 urlencode(base64_encode($path)),
+                 md5($path . $secret)
+  );
+}
+
+/**
+ * Check whether a user is logged in via UNI•Login.
+ */
+function ding_unilogin_authenticated($account = NULL) {
+  global $user;
+  if (!$account) {
+    $account = $user;
+  }
+
+  if (isset($account->data['unilogin']) && $account->data['unilogin']) {
+    return TRUE;
+  }
+  return FALSE;
+}
+
+/**
+ * Attemp to log in UNI•Login user.
+ *
+ * @param string $username
+ *   The UNI•Login user name.
+ *
+ * @return bool
+ *   Whether the user was logged in.
+ */
+function _ding_unilogin_login($username) {
+  $auth_service = _ding_unilogin_wservice('auth');
+  $access = FALSE;
+
+  try {
+    // Check that the user has license to access us.
+    $access = $auth_service->hasLicense(
+      $username,
+      variable_get('ding_unilogin_auth_context', ''),
+      variable_get('ding_unilogin_auth_project', '')
+    );
+
+    // Look up the retailer id we should use for the user.
+    if ($access) {
+      $info_service = _ding_unilogin_wservice('info');
+      $municipality = $info_service->getMunicipality($username);
+      $municipalities = variable_get('ding_unilogin_municipality_mapping', array());
+      if (isset($municipalities[$municipality])) {
+        $publizon_id = $municipalities[$municipality];
+
+        // Check that the retailer_id is one of the configured libraries.
+        $publizon_libraries = variable_get('publizon_libraries', array());
+        if (!isset($publizon_libraries[(int) $publizon_id])) {
+          $access = FALSE;
+        }
+      }
+      else {
+        $access = FALSE;
+      }
+    }
+  }
+  catch (Exception $e) {
+    drupal_set_message(t('Sorry, but there was problems communicating with UNI•Login. Please try again later.'));
+    watchdog_exception('ding_unilogin', $e, NULL, array(), WATCHDOG_ERROR);
+    return FALSE;
+  }
+
+  if (!$access) {
+    drupal_set_message(t('Sorry, but that UNI•Login account has not been granted access'), 'error');
+    watchdog('ding_unilogin', 'Access denied for UNI•Login user %user', array('%user' => $username), WATCHDOG_NOTICE);
+    return FALSE;
+  }
+
+  $auth_name = ding_user_default_authname($username);
+  $auth_res = array(
+    'success' => TRUE,
+    'creds' => array(
+      'name' => $username,
+      // Random password. Shouldn't be used, but we'll make it un-guessable
+      // for good measure, in case someone messes up.
+      'pass' => md5(drupal_random_bytes(32) . 'no_password_for_the_wicked'),
+    ),
+    'user' => array(
+      'mail' => '',
+      'blocked' => FALSE,
+      'data' => array(
+        'display_name' => '',
+      ),
+    ),
+  );
+
+  /*
+   * We call two private ding_user functions to do the heavy lifting. Sadly,
+   * nobody has re-factored ding_user to have a simple login function, so we
+   * have to duplicate its work here.
+   */
```

Comment:
No it's not. ding_wayf simply submits the user_login form, sidestepping the issue.

I'd love to refactor ding_user, ding_wayf and this to something much prettier, but haven't gotten the time.